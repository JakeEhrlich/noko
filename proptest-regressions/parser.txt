# Seeds for failure cases proptest has generated in the past. It is
# automatically read and these particular cases re-run before any
# novel cases are generated.
#
# It is recommended to check this file in to source control so that
# everyone who runs the test benefits from these saved cases.
cc 4b2fb53ee6f871f61c8f8cddf604c670431eea8aa1ab07b466426f9cfa303b38 # shrinks to expr = StringLiteral("\"\"")
cc 9dc081a15abc5c6eaf94eaa66b537fbb36258e9401b7dd00c73c58c422fbba39 # shrinks to expr = Call([Call([]), Var("$")])
