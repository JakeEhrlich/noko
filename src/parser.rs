use chumsky::error::SimpleReason;
use chumsky::prelude::*;
use miette::Diagnostic;
use miette::LabeledSpan;
use miette::SourceSpan;
use thiserror::Error;

#[derive(Debug)]
pub enum Expr {
    Var(String),
    Call(Vec<Expr>),
    StringLiteral(String),
    ColorLiteral(String),
}

const fn assci_char_set(set: &'static str) -> [bool; 256] {
    let mut out = [false; 256];
    let mut idx = 0;
    let bset = set.as_bytes();
    while idx < bset.len() {
        out[bset[idx] as usize] = true;
        idx += 1;
    }
    out
}

fn in_assci_set(c: char, set: &[bool; 256]) -> bool {
    let true = c.is_ascii()
    else {
      return false;
    };
    set[u32::from(c) as usize]
}

fn is_leading_var_char(c: &char) -> bool {
    const CSET: [bool; 256] = assci_char_set("!$%&*+-./:<=>?@^_~");
    c.is_alphabetic() || in_assci_set(*c, &CSET)
}

fn is_var_char(c: &char) -> bool {
    is_leading_var_char(c) || c.is_ascii_digit()
}

fn varible() -> impl Parser<char, Expr, Error = Simple<char>> {
    filter(is_leading_var_char)
        .chain(filter(is_var_char).repeated())
        .collect()
        .map(Expr::Var)
        .padded()
        .padded_by(comment())
}

fn comment() -> impl Parser<char, (), Error = Simple<char>> + Clone {
    just("//")
        .then(take_until(text::newline()))
        .padded()
        .repeated()
        .ignored()
}

fn string_literal() -> impl Parser<char, Expr, Error = Simple<char>> {
    just("\"")
        .ignore_then(take_until(just("\"").ignored()))
        .map(|t| t.0)
        .collect()
        .map(Expr::StringLiteral)
        .padded()
        .padded_by(comment())
}

fn is_hex_digit(c: &char) -> bool {
    c.is_ascii_hexdigit()
}

fn hex_color_literal() -> impl Parser<char, Expr, Error = Simple<char>> {
    just("#")
        .ignore_then(filter(is_hex_digit).repeated().exactly(6))
        .collect()
        .map(Expr::ColorLiteral)
        .padded()
        .padded_by(comment())
}

pub fn expr() -> impl Parser<char, Vec<Expr>, Error = Simple<char>> {
    recursive(|e| {
        choice((
            varible(),
            string_literal(),
            hex_color_literal(),
            e.delimited_by(just('('), just(')')).map(Expr::Call),
        ))
        .repeated()
        .at_least(1)
    })
    .padded()
    .padded_by(comment())
}

pub fn to_parse_error(source: &str, err: &Simple<char>) -> ParseError {
    match err.reason() {
        SimpleReason::Unexpected => {
            let mut lbls: Vec<_> = err.label().into_iter().map(|_| err.span().into()).collect();
            if lbls.is_empty() && err.found().is_none() {
                lbls = vec![(source.len() - 1..source.len()).into()];
            }
            ParseError::Unexpected {
                found: err.found().copied(),
                expected: err.expected().flatten().cloned().collect(),
                label: lbls.into_iter().next(),
            }
        }
        SimpleReason::Unclosed { span, .. } => ParseError::Unmatched {
            begin: LabeledSpan::at(span.clone(), "Opening paren here"),
            end: LabeledSpan::at(err.span(), "mismatch here"),
        },
        _ => {
            ParseError::Generic {
                msg: err.to_string(),
            }
        }
    }
}

#[derive(Error, Debug, Diagnostic)]
pub enum ParseError {
    #[error("Unexpected Syntax: expected one of {expected:?}, found {found:?}")]
    #[diagnostic(code(parser::unexpected_input))]
    Unexpected {
        found: Option<char>,
        expected: Vec<char>,
        #[label]
        label: Option<SourceSpan>,
    },
    #[error("Unmatched Paren")]
    #[diagnostic(code(parser::unmatched_paren))]
    Unmatched {
        begin: LabeledSpan,
        end: LabeledSpan,
    },

    #[error("Parse Error: {msg}")]
    #[diagnostic(code(parser::generic_error))]
    Generic {
        msg: String,
    },
}

#[derive(Debug, Error, Diagnostic)]
#[error("Parsing Errors")]
#[diagnostic(code(parse_error))]
pub struct ParseErrors {
    #[source_code]
    pub source_code: String,

    #[related]
    pub all: Vec<ParseError>,
}

pub fn parse(source: &str) -> Result<Vec<Expr>, ParseErrors> {
    expr()
        .parse(source)
        .map_err(|errs| ParseErrors {
            source_code: source.into(),
            all: errs.iter().map(|err| to_parse_error(source, err)).collect()})
}
