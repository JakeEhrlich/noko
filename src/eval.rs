use crate::parser::Expr;
use data_encoding::HEXLOWER_PERMISSIVE;
use levenshtein::levenshtein;
use miette::Diagnostic;
use owo_colors::Style;
use std::collections::HashMap;
use std::ops::Range;
use std::sync::Arc;
use thiserror::Error;

pub type EvalFunc = &'static fn(&[Value]) -> Value;

// This is a high level byte code for a stack
// machine. We might optimize this later into
// something even more efficent if needed.
enum ByteCode<'a> {
    Call(EvalFunc, usize),
    PushConstant(&'a Value),
    PushLiteral(Value),
}

// TODO: Add time, duration, size, maybe other things
pub enum Value {
    Color { r: u8, g: u8, b: u8 },
    Style(Arc<Style>),
    String(Arc<String>),
    Integer(i64),
}

#[derive(Error, Debug, Diagnostic)]
pub enum EvalError {
    #[error("Constant is not registerd")]
    #[diagnostic(code(eval::unregisterd_constant))]
    UnregistredConstant {
        #[help]
        help: Option<String>,

        #[label("constant seen here")]
        constant: Range<usize>,
    },

    #[error("Invalid hex color code")]
    #[diagnostic(code(eval::invalid_hex))]
    InvalidHex {
        #[label("here")]
        hex: Range<usize>,
    },

    #[error("Empty parens are not valid")]
    #[diagnostic(code(eval::empty_parens))]
    EmptyParens {
        #[label("here")]
        parens: Range<usize>,
    },

    #[error("Attempt to call a literal. Only registerd functions can be called.")]
    #[diagnostic(code(eval::call_literal))]
    CallLiteral {
        #[label("here")]
        literal: Range<usize>,
    },

    #[error("Function is not registerd")]
    #[diagnostic(code(eval::unregisterd_function))]
    UnregistredFunction {
        #[help]
        help: Option<String>,

        #[label("function seen here")]
        function: Range<usize>,
    },
}

fn find_closest_match<T>(found: &String, dict: &HashMap<String, T>) -> Option<(String, usize)> {
    dict.keys()
        .map(|s| (s, levenshtein(s, found)))
        .min_by_key(|p| p.1)
        .map(|p| (p.0.clone(), p.1))
}

fn handle_unreg_constant(
    found: &String,
    constants: &HashMap<String, Value>,
    reg: &HashMap<String, EvalFunc>,
) -> Option<String> {
    // First handle mismatched name spaces
    if reg.get(found).is_some() {
        // TODO: Add url to documentation on this
        // TODO: Add an error code help command that can be called
        return Some(format!(
            "{} is a function, not a constant, did you mean to call it?",
            found
        ));
    }

    // Now that we know there isn't a match anywhere lest find the closest match
    let Some((closest_const, cdist)) = find_closest_match(found, constants) else {
        return None;
    };

    // We should also look through functions to see if we find a match there
    let Some((closest_func, fdist)) = find_closest_match(found, reg) else {
        // If there's no best function we can return early
        if (cdist as f64) < 0.33 * (found.len() as f64) {
            return Some(format!("did you mean {} perhaps?", closest_const));
        } else {
            return None;
        }
    };

    if fdist < cdist {
        if (fdist as f64) < 0.33 * (found.len() as f64) {
            return Some(format!(
                "did you mean {} perhaps? note that {} is a function not a constant",
                closest_func, closest_func
            ));
        } else {
            return None;
        }
    } else {
        if (cdist as f64) < 0.33 * (found.len() as f64) {
            return Some(format!("did you mean {} perhaps?", closest_const));
        } else {
            return None;
        }
    }
}

fn handle_unreg_function(
    found: &String,
    constants: &HashMap<String, Value>,
    reg: &HashMap<String, EvalFunc>,
) -> Option<String> {
    // First handle mismatched name spaces
    if constants.get(found).is_some() {
        // TODO: Add url to documentation on this
        // TODO: Add an error code help command that can be called
        return Some(format!(
            "{} is a constant, not a function and cannot be called",
            found
        ));
    }

    // Now that we know there isn't a match anywhere lest find the closest match
    let Some((closest_func, fdist)) = find_closest_match(found, reg) else {
        return None;
    };

    // We should also look through functions to see if we find a match there
    let Some((closest_const, cdist)) = find_closest_match(found, constants) else {
        // If there's no best function we can return early
        if (fdist as f64) < 0.33 * (found.len() as f64) {
            return Some(format!("did you mean {} perhaps?", closest_func));
        } else {
            return None;
        }
    };

    if cdist < fdist {
        if (cdist as f64) < 0.33 * (found.len() as f64) {
            return Some(format!(
                "did you mean {} perhaps? note that {} is a constant not a function",
                closest_const, closest_const
            ));
        } else {
            return None;
        }
    } else {
        if (fdist as f64) < 0.33 * (found.len() as f64) {
            return Some(format!("did you mean {} perhaps?", closest_func));
        } else {
            return None;
        }
    }
}

fn to_byte_code<'a>(
    constants: &'a HashMap<String, Value>,
    reg: &HashMap<String, EvalFunc>,
    expr: &Expr,
    out: &mut Vec<ByteCode<'a>>,
) -> Result<(), EvalError> {
    match expr {
        Expr::Var(name, span) => {
            let Some(value) = constants.get(name) else {
                return Err(EvalError::UnregistredConstant {
                    help: handle_unreg_constant(name, constants, reg),
                    constant: span.clone(),
                });
            };
            out.push(ByteCode::PushConstant(value));
        }
        Expr::Call(vec, span) => {
            if vec.is_empty() {
                return Err(EvalError::EmptyParens {
                    parens: span.clone(),
                });
            }
            let Expr::Var(func_name, func_span) = &vec[0] else {
                return Err(EvalError::CallLiteral {
                    literal: vec[0].span().clone(),
                });
            };
            let Some(eval_func) = reg.get(func_name) else {
                return Err(EvalError::UnregistredFunction {
                    help: handle_unreg_function(func_name, constants, reg),
                    function: func_span.clone(),
                });
            };
            // Push everything.
            for expr in &vec[1..] {
                // TODO: Return multiple errors here
                to_byte_code(constants, reg, expr, out)?;
            }
            out.push(ByteCode::Call(*eval_func, vec.len() - 1));
        }
        Expr::StringLiteral(str, _) => {
            out.push(ByteCode::PushLiteral(Value::String(Arc::new(str.clone()))))
        }
        Expr::ColorLiteral(color, span) => {
            let data = HEXLOWER_PERMISSIVE
                .decode(color.as_bytes())
                .map_err(|_| EvalError::InvalidHex { hex: span.clone() })?;
            if data.len() != 3 {
                return Err(EvalError::InvalidHex { hex: span.clone() });
            }
            let color = Value::Color {
                r: data[0],
                g: data[1],
                b: data[2],
            };
            out.push(ByteCode::PushLiteral(color));
        }
    };
    Ok(())
}
