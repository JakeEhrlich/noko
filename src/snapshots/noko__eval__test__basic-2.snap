---
source: src/eval.rs
expression: "test(\"(head a b a)\")"
---
Ok(
    [
        pushc String("a"),
        pushc Color { r: 33, g: 33, b: 33 },
        pushc String("a"),
        call <func> 3,
    ],
)
